src/Bounds.cpp

```diff
@@ -1046,6 +1047,86 @@ bool box_contains(const Box &outer, const Box &inner) {
     return is_one(simplify(condition));
 }
 
+class FindInnermostVar : public IRVisitor {
+public:
+    const map<string, int> &vars_depth;
+    string innermost_var;
+
+    FindInnermostVar(const map<string, int> &vars_depth)
+        : vars_depth(vars_depth) {}
+
+private:
+    using IRVisitor::visit;
+    int innermost_depth = -1;
+
+    void visit(const Variable *op) {
+        const auto &iter = vars_depth.find(op->name);
+        if ((iter != vars_depth.end()) && (iter->second > innermost_depth)) {
+            innermost_var = iter->first;
+            innermost_depth = iter->second;
+        }
+    }
+};
+
+// Place innermost vars in an IfThenElse's condition as far to the left as possible.
+class SolveIfThenElse : public IRMutator {
+    // Map of variable names and their depths. Higher depth indicates
+    // variable defined more innermost.
+    map<string, int> vars_depth;
+    int depth = -1;
+
+    using IRMutator::visit;
+
+    int push_var(const string &var) {
```

Comment:
I think this would be simpler with a Scope<int>. push_var would just be:

vars_depth.push(var, vars_depth.size());

It would shadow any previously-pushed var of the same name until it gets popped.